### DISCUSSION TITLE
Publish Android Library

### DISCUSSION THREAD

**Discussion by benjohnde** [zxing-cpp/zxing-cpp#636] [Ideas]
@axxel would it be okay if I invest some work to publish the Android library as well? I did so with a few of our little KMP libraries (cf. https://github.com/Liftric/KVault). We can automate all the things, so you won't have any additional work. For most people the release / artifact approach is probably working though but not super clean IMHO.

**Comment by axxel** [comment]
You mean Maven and/or Sonatype packages? Sounds like a good idea, judging how may people access the python wrapper via PyPi and how few seem to get the artifacts stuff or simply build themselves.

**Reply by axxel** [reply]
After all the API improvements discussed below, now back to the original discussion about publishing on mavenCentral:

@markusfisch you've prepared this to some extend in https://github.com/zxing-cpp/zxing-cpp/commit/0b4addbfbbf63016b03a2a32bed93e081cae3fbb. Do you have a clear understanding what is still missing to actually get it published? I had a short look at the [requirements](https://maven.apache.org/repository/guide-central-repository-upload.html) and there might still be some work to be done (e.g. POM / javadoc)?

**Reply by markusfisch** [reply]
> Do you have a clear understanding what is still missing to actually get it published?

Only in theory. Unfortunately, there are many conflicting or outdated guides, and the [offical documention](https://developer.android.com/build/publish-library/upload-library) is way too sparse.

From reading this [guide](https://medium.com/@vivekvashistha/how-to-publish-android-libraries-on-maven-central-in-2023-a96e3c327008) (which most probably does too much), my understanding is that you would need to

1. register the Group ID on Maven by [raising a Jira Ticket](https://central.sonatype.org/publish/publish-guide/#deploying-to-ossrh-with-gradle-introduction).
2. complete `gradle.properties` with meta data for the POM (which is automatically generated by the [maven-publish](https://docs.gradle.org/current/userguide/publishing_maven.html) plugin) like it's shown in the guide (or check this real world [sample](https://github.com/chrisbanes/accompanist/blob/main/gradle.properties#L38C7-L38C7))
3. add a publication for "sonatype" in the [publishing](https://github.com/zxing-cpp/zxing-cpp/blob/0b4addbfbbf63016b03a2a32bed93e081cae3fbb/wrappers/android/zxingcpp/build.gradle.kts#L56) block like it's done in `publish-package.gradle` in the guide
4. configure Maven publishing credentials and signing keys and reference them in (your private global) `gradle.properties`
5. publish the release by running `./gradlew publishReleasePublicationToSonatypeRepository`
6. manually approve the staging on [Nexus Dashboard](https://s01.oss.sonatype.org/) with the Jira Credentials

Unfortunately, I can't do any of this because that would mean taking ownership of these things.

Sorry for not being able to help more with this needlessly complex process.

**Reply by axxel** [reply]
Thanks for all this info and research. I spend the better part of a day now trying to find my way through the config/setup jungle (landed on a few other more kotlin DSL specific guides and ended up close to benjohnde's KVault sample). I have the [Group ID](https://issues.sonatype.org/browse/OSSRH-97089) and the GPG keys setup and ready in GA secrets to be used.

Where I currently got stuck in was the next round of naming questions with the different constraints imposed by Maven/sonartype.org, Java package naming and our precedent in the form of the python and cocoapods packages. To my understanding the `Group ID` _has_ to be `io.github.zxing-cpp` while the java/kotlin package name _can't_ be `io.github.zxing-cpp` because of the '-'. According to [this official info](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html) one is advised to replace a '-' with a '_'.

And just to make it more fun there are the following places where some form of the name is expected:
1. `settings.gradle.kts`: `rootProject.name`
2. `build.gradle.kts`: `android/namespace`, `group`, `MavenPublication/artifactId`,  `[...]/pom/name`.

As far as I understand it, the Kotlin object `ZXingCpp` can't possibly have a '-' in the name and using PascalCase for class/singleton names is the canonical form for all Java/Kotlin, right?

Questions to @benjohnde and @markusfisch:
1. did I overlook something in the above summary?
2. is there any better option than replacing the '-' with a '_' in the Java package name?
3. do you have any input regarding the mentioned "more fun" names above?
4. which of all the names is the one going to end up in the `dependencies` section of a client application?

EDIT:
After some more reading my current understanding is:
* snoatype.org groupID must be `io.github.zxing-cpp` (even if a '-' can not be used in the java package name)
* java package name can be anything (e.g. stay `com.zxingcpp` or become `io.github.zxing_cpp` or `io.github.zxingcpp` or simply `zxingcpp`). It seems to be rather uncommon that people use the `io.github` prefix and prefer stuff like the traditional `com.` variant even when their project page is on github and thy don't have any domain.
* `android/namespace` and `android/applicationID` are supposed to be equal and the same as the package name but don't have to be and since they seem to be related mostly to identifying an app in the play store, I don't even know what they mean in the context of a module/library? In fact, I found a few libraries that don't seem to use that property, which looks like the best choice as we would not have to worry about it then? sigh.

**Reply by markusfisch** [reply]
Sorry for all the trouble 😬 

A hyphen in the Group ID is okay, and while a Group ID _can_ match the package name of a library, it does not have to.

The Group ID is only used for dependency resolution. The most common [notation](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/dsl/DependencyHandler.html) for the `implementation` statement in  `dependencies` is "group:name:version", where group is the `groupId` and name the `artifactId` from the POM (which is automatically generated from the [publishing](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/build.gradle.kts#L58) section). Both are just names and can contain hyphens.

Ideally, the only place where `groupId` and `artifactId` need to be specified are in `zxingcpp/build.gradle` [here](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/build.gradle.kts#L61). 

> android/namespace and android/applicationID are supposed to be equal and the same as the package…

Android libraries [need to specify](https://developer.android.com/build/publish-library/prep-lib-release#choose-namespace) a `namespace` because it is used for resource compilation. This really should match the Java package name.

The library project (configured by [zxingcpp/build.gradle.kts](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/build.gradle.kts)) does not have an `applicationId`, as that is a property of applications only. So this is irrelevant in this context.

**Reply by markusfisch** [reply]
So, looking at [zxingcpp/build.gradle](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/build.gradle.kts#L61), `groupId` should be "io.github.zxing-cpp" now.

With this in place, the `implementation` statement (of a consuming app) would have to look like this:
```
dependencies {
    …
    implementation 'io.github.zxing-cpp:zxingcpp:2.1.0'
}
```

And I think you'd need a `repositories`/`maven` block for "sonatype" in [publishing](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/build.gradle.kts#L58) like it is in the [guide](https://gist.github.com/vkvashistha/ec4bcb437724718a3c04ab0eee447b87#file-publish-package-gradle-L91).

**Reply by axxel** [reply]
Thanks again for your time. So we agree that `groupId` is settled. That leaves:

**artifactId**
I saw that `artifactId` defaults to `project.name` which (I assume) defaults to `rootProject.name` from [here](https://github.com/zxing-cpp/zxing-cpp/blob/475d61f85a45f4544bb6b8315087820257babd11/wrappers/android/settings.gradle.kts#L19). The best choice for `artifactId` is unclear to me. I saw that e.g. upstream ZXing uses `core` (which might also be in line with our cocoapods `Core` spec, even though there this only refers to the c++ part while here it would mean the wrapper as a whole). I also saw multiple occasions where people seemed to have used the term `android`. As of now I would consider using `core` or `android` or `zxing-cpp`. Thoughts?

**Java package name**
`io.gihub.zxing_cpp` or `io.github.zxingcpp` or `com.zxingcpp` (_all_ of which would be inconsistent with the `groupId`. while the '_' version would apparently be the most canonical choice, I'd lean towards `zxingcpp` just to not introduce yet another way of spelling the name of the project :-/). If we were to consider making the demo-app public or put it in the play store, we'd need a different name for that, right? like `io.gihub.zxingcpp.app` would this then suggest to add something like `.core` (see above) to the package name of the lib? One would then `import io.github.zxingcpp.core.ZXingCpp` in a client source file?

(Forgot to mention @okarmazin earlier who might also have some opinion.)

**Reply by okarmazin** [reply]
Hello, I do maintain a library that I publish to Maven Central. Hopefully I can be of service here and sorry for not responding earlier.

I use this Gradle plugin: https://github.com/vanniktech/gradle-maven-publish-plugin

The final publication setup looks like this: https://github.com/okarmazin/escpos4k/blob/5dd502d0d46b22572dcfe80ce7486486be90754e/escpos4k/build.gradle.kts#L16

I could prepare the initial build file setup if you want - and after the GPG key is [set up](https://vanniktech.github.io/gradle-maven-publish-plugin/central/#secrets) on the CI in the correct environment variables it should be smooth sailing.

**Reply by markusfisch** [reply]
> artifactId

The `project.name` is used to identify a project (in a [multi-project](https://docs.gradle.org/current/userguide/multi_project_builds.html#sec:creating_multi_project_builds) environment) in a Gradle build. `rootProject.name` is the name of the root project of a  multi-project environment.

The `artifactId` is used to identify an artifact (e.g. library), in the given group, in Maven dependency resolution.

They _can_ be identical, which usually makes sense, but there is no hard connection between these two domains.

I would suggest `zxingcpp` because the artifact is an Android library that contains the wrapper and the native library (core). `zxing-cpp` is a bit more cumbersome, and also breaks with the current Java package name, which is `com.zxingcpp`.

An alternative could be `android` if we are planning to release a pure Java/JVM wrapper some day. Then, there would have to be two artifacts: one for the Android library (the `aar`), and one for the JVM (which would be a `jar`).

> Java package name

For me, the current package name, `com.zxingcpp`, is the most convenient. It's short and simple and free from special characters that can cause problems here and there.

> If we were to consider making the demo-app public or put it in the play store, we'd need a different name for that, right? 

No, the demo app can have any package name (or rather applicationId as it's called then) that is not already taken. There is no connection to the library, GitHub or the Maven repositories.

> One would then import io.github.zxingcpp.core.ZXingCpp in a client source file?

Yes, if we would change the package name to `io.github.zxingcpp.core`. But I don't see much value in the `.core` part. If there would be more Classes, and it would be desirable to separate them in different package names, then I would add something like `.core`.

If you're going with `android` for the artifactId, I would suggest to include `.android` in the package name as well (something like `io.github.android.zxingcpp`). This would communicate that this package can only be used with the Android SDK.

**Reply by axxel** [reply]
Then I kind of like the idea to use `android` which would then settle the `artifactId` / gradle naming: `implementation("io.github.zxing-cpp:android:2.1.0")`

Final question: package name:

In terms of `core`: I thought about introducing the idea of splitting the reader and writer part in two modules which would actually make some sens from a binary size minimization perspective but did not want to make it even more complicated.

I assume the place of the string `android` in `io.github.android.zxingcpp` was a typo and it was supposed to come at the end?

I assume there would be no point in including both a aar and jar version into one project, and if `.core` has not much value, then I don't see why adding `.android` to the class path brings much value. The user will see that name in the `implementation` statement, so that information is conveyed then.

If the class path can't be consistent with `groupId` and the `com.` prefix is arbitrary why then not being creative? Quick poll for my current favorites:
👍 = `import zxing.cpp.ZXingCpp`
👎 = `import io.github.zxingcpp.ZXingCpp`

@okarmazin thanks for the hint at your project but the open questions regarding naming are independent of that and blocking any further progress towards a specific gradle implementation.

**Reply by markusfisch** [reply]
> Then I kind of like the idea to use android which would then settle the artifactId / gradle naming:

Looks good to me!

> I thought about introducing the idea of splitting the reader and writer part in two modules which would actually make some sens from a binary size minimization perspective but did not want to make it even more complicated.

In theory a good idea - but both artifacts would have to contain the very same native library, so this would be very undesirable for apps that use both.

> I assume the place of the string android in io.github.android.zxingcpp was a typo and it was supposed to come at the end?

No, because `android` would be a more general filter than the package itself (which is only available on Android).

> I assume there would be no point in including both a aar and jar version into one project

Most likely not, because the JVM version would use a [BufferedImage](https://docs.oracle.com/javase/8/docs/api/java/awt/image/BufferedImage.html) instead of a [Bitmap](https://developer.android.com/reference/android/graphics/Bitmap), for example. Neither one exist on the other platform.

> and if .core has not much value, then I don't see why adding .android to the class path brings much value. The user will see that name in the implementation statement, so that information is conveyed then.

True, this is probably more a matter of taste. It's more accurate, and if you ever going to exchange code between Android and the JVM (backend), it can be desirable to immediately know what package is platform specific. But YMMV, of course, and it's really not that important.

> If the class path can't be consistent with groupId and the com. prefix is arbitrary why then not being creative?

I like `import io.github.zxingcpp.ZXingCpp` better, because `github.io` is a valid URL, and you could (in theory) put up a page for ZxingCpp there (some day).

**Reply by axxel** [reply]
> both artifacts would have to contain the very same native library

That would of course make no sense at all and is not what I meant. But the current choice of naming would later allow for that still.

> I like import io.github.zxingcpp.ZXingCpp better, because github.io is a valid URL

True but if I look at [that code](https://github.com/zxing-cpp/zxing-cpp/blob/475d61f85a45f4544bb6b8315087820257babd11/wrappers/android/app/src/main/java/com/example/zxingcppdemo/MainActivity.kt#L19) then having the class path being a valid URL seems pretty outdated. I understand that this whole domain name thing was an easy solution to leverage an existing globally managed namespace which is not universally used like that (anymore) and I see a value in using that as a pointer to webpage (that does not exist a.t.m. anyway). But thanks for your opinion. I did ask for it, it is just that this particular argument is not very convincing to me. I was mostly afraid I would overlook something serious with the "creative" approach.

**Reply by markusfisch** [reply]
Well, the platform is really something general. There was always the [java](https://github.com/zxing-cpp/zxing-cpp/blob/475d61f85a45f4544bb6b8315087820257babd11/wrappers/android/app/src/main/java/com/example/zxingcppdemo/MainActivity.kt#L56) namespace, which of course is not a URL either.

To me there's some value in being able to deduce a web address from a package name. But that is of course a matter of opinion, and there is really nothing wrong with `zxing.cpp.ZXingCpp`.

**Reply by axxel** [reply]
After reading more on this subject I came to the conclusion that the `zxing.cpp` variant is actually a bit silly and as with `zxing_cpp`, it would be yet another new name. So `zxingcpp` it is. As simple as that. This name is unique enough for me and adding something like `com.` in front of it without actually owning that domain adds no value whatsoever.

I'll push something soonish.

**Reply by axxel** [reply]
I pushed the changes, added a publish-android action and here is the "2.1.0-SNAPSHOT" version uploaded: https://s01.oss.sonatype.org/content/repositories/snapshots/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/

Can someone find out if this now actually "works"? Are any of these [two](https://github.com/zxing-cpp/zxing-cpp/blob/26f3a3d8581d76562c6a94fd0456b03e7484926a/wrappers/android/zxingcpp/build.gradle.kts#L21) [lines](https://github.com/zxing-cpp/zxing-cpp/blob/26f3a3d8581d76562c6a94fd0456b03e7484926a/wrappers/android/zxingcpp/build.gradle.kts#L28) an issue when one want's to have a 'properly' released android library?

**Reply by markusfisch** [reply]
> Can someone find out if this now actually "works"?

Hm, unless I did something silly, it doesn't work (yet):

```
$ ./gradlew buildDebug 
> Task :app:mapDebugSourceSetPaths FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mapDebugSourceSetPaths'.
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Could not find io.github.zxing-cpp:android:2.1.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/android-2.1.0-SNAPSHOT.pom
       - https://dl.google.com/dl/android/maven2/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/maven-metadata.xml
       - https://dl.google.com/dl/android/maven2/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/android-2.1.0-SNAPSHOT.pom
       - https://s01.oss.sonatype.org/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/maven-metadata.xml
       - https://s01.oss.sonatype.org/io/github/zxing-cpp/android/2.1.0-SNAPSHOT/android-2.1.0-SNAPSHOT.pom
     Required by:
         project :app
```

Maybe Sonatype hasn't published this officially?

Here's a patch file to make the sample app in `wrappers/android` use the dependency, in case you want to try it yourself:

[include_zxing_from_sonatype.patch](https://github.com/zxing-cpp/zxing-cpp/files/13525889/include_zxing_from_sonatype.patch)

> abiFilters += "armeabi-v7a"

Depends. If you want to support (Android apps on) Chromebooks too, then the `abiFilters` would be a problem. Google recommends supporting Chromebooks, of course, but you can always add support in a later version.

> arguments("-DCMAKE_BUILD_TYPE=RelWithDebInfo")

Including the debug info in the native library is actually recommended by Google Play.

**Reply by axxel** [reply]
Thanks for your input as always. As you can see in #655, I managed to make it work for me based on your patch with an updated maven URL.

**Reply by markusfisch** [reply]
Happy to help! And sorry for the wrong URL 🙈

**Comment by axxel** [comment]
I'm hijacking this discussion a little bit for the sake of discussing other changes regarding the Android library in the context of a new release: I asked @markusfisch whether he would be interested in bringing some of his modifications from his own [android wrapper](https://github.com/markusfisch/zxing-cpp/blob/master/wrappers/aar/zxingcpp/src/main/java/de/markusfisch/android/zxingcpp/ZxingCpp.kt) flavor to upstream. My idea was to use this momentum we currently have regarding the iOS wrapper and bring the Android side along, trying to make them feel equally strong/usable. This will result in breaking changes but now seems to be the best time for that in my opinion. There are 3 areas that I'd like to adjust:
1) make the `read` call return a list of codes instead of just one
2) extend the `Options` struct with more `DecodeHints` (although not necessarily all of them)
3) extend the `Result` struct to carry more of the c++ Result members (no `version` and I'm not convinced of the `GTIN` substruct on the iOS side)

I currently would refrain from adding write support as long as the underling interface is still only an uncommited rough sketch on my hard drive and no one is demanding it yet. About the term `Options`: I chose that one as I got the impression that this seemed like a standard pattern on Android for this kind of data class and wanted to make it feel 'native'. There is obviously an argument to be made to call it `DecodeHints` as well.

Comments welcome (e.g. from @benjohnde @markusfisch @okarmazin).

**Reply by benjohnde** [reply]
Hi there, I am currently pretty busy, sry for the late reply (applies to other discussions as well)!

I guess most people will use `read` capabilities and not so much the `write` part. For me your suggestions are in sound with improving the overall library usability. If all three implementations are streamlined (Android to iOS to C++) and it is published on Maven for easier usage, more people will probably use this port instead of the original ZXing.

So, the breaking changes are worth it! Thanks @markusfisch for taking the initiative to push the Android wrapper and of course @axxel for the library as a whole! It makes our lives easier :)

**Reply by markusfisch** [reply]
I could create PRs for each of these 3 areas, and this is most probably the best thing to do.

But I think one of the biggest issues with the official wrapper is that it's not published on Maven (or any repository for that matter). This makes it a lot less approachable than the Java implemenation, which is available through Maven. 

Also, distributing the AAR directly doesn't provide a user with any information regarding the identity, version, or dependencies of a library. Therefore, Google strongly [recommends ](https://developer.android.com/build/publish-library/upload-library#distribution) to publish libraries using repositories rather than manually distributing AAR files.

So, like @benjohnde, I think this is probably the most important thing to push ZXingCpp for Android.

That said, it would be wise to sort some things out before publishing, of course, and this includes the 3 areas you mentioned. But there are also a few (formal) things that I would suggest changing:

1. Rename `BarcodeReader` to `ZXingCpp` for two reasons:
* perspectively we are going to include write support and then, `BarcodeReader` wouldn't be a good fit. Of course, we could also add `BarcodeWriter`, but then there is `BarcodeReader.Formats`, which we also want to use when writing a barcode.
* reading `ZXingCpp.read()` (in client code) makes it immediately clear to anyone that this is a function of ZXingCpp, whereas `BarcodeReader.read()` could be something of the client too.

2. Make `options` an argument to `read()` instead of a field of `BarcodeReader`. `options` is it's own object anyway, and by separating it from the object, it's clear where and when it is used. This makes the API a little bit more "pure", meaning more clear about side effects. Also, there's already `read(bitmap: Bitmap, options: Options, …)` so this would make the API more consistent as well.

3. Make `BarcodeReader` an `object` instead of a `public class`. Using `object` is Kotlin's way of [defining a Singleton](https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview). With `options` being an argument, `BarcodeReader` can be a Singleton. Again, this makes the API more clear to the user about possible side effects, and then the user would not have to instantiate `BarcodeReader` but only `Options`.

---

> I currently would refrain from adding write support as long as the underling interface is still only an uncommited rough sketch on my hard drive and no one is demanding it yet.

Very well, but that would mean that the Android wrapper is not "equally strong/usable" 😉 If this library should be a drop-in replacement for the Java implemenation, it needs to have write support.

I know you're in the middle of refactoring this, but this is already in use in the native library, as well as in the iOS wrapper, and in my Android wrapper. So at least we are using it, and it's really important to us. There are a lot applications for recreating a scanned barcode, for example, and I think the current solution is good enough for that. Of course, there will be improvements, even breaking changes, but this is not a good reason to not make this available now.

> About the term Options: I chose that one as I got the impression that this seemed like a standard pattern on Android for this kind of data class and wanted to make it feel 'native'.

I would be very much in favor of `DecodeHints`, as this is the native name and also the term we used in the iOS wrapper. Yes, there are things like `BitmapFactory.Options`, but Options is a very general term, and here we can name things more precisely. These aren't just any options, but decoding hints, so it's better to make clear what these options are.

---

Then, just to mention this again so that nobody can say that nobody said anything 😉: The Android wrapper should not use `ImageProxy` as this requires including CameraX just for the sole reason of making `BarcoderReader.read()` conveniently having fewer arguments at the cost of forcing everyone to use CameraX, no matter if the app is using it or not. In my opinion, this only creates unnecessary hurdles.

**Reply by axxel** [reply]
Thanks @markusfisch for your valuable input. I agree with you on everything except the write-support (I understand you rely on it but I have the feeling you would keep your own fork of the wrapper anyway?) and the removal of the CameraX dependency. I'd be happy to merge a PR (single one is fine for me) that deals with all the discussed improvements.

**Reply by axxel** [reply]
@markusfisch do you know of a way to have the `ImageProxy` variant of `read` be somehow dependent on whether CameraX is available or not? Something along the lines of the `if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {` check? Wondering whether there was a way to not force the depency on CameraX but still provide the convenient `read(ImageProxy)` usage if the client uses CameraX.

**Reply by markusfisch** [reply]
@axxel Well, yes, this is possible, but we would have to change the argument type (from `ImageProxy`) to `Object` and use [reflection](https://www.oracle.com/technical-resources/articles/java/javareflection.html) to get the relevant fields from the given (`ImageProxy`) object.

`ImageProxy` can't be used in the library directly then, because it wouldn't be there at compile time without the CameraX library.

But this would probably be quite a hack 😉

**Comment by axxel** [comment]
publishing on mavenCentral is <insert swear word of your choice>. But there it is: https://central.sonatype.com/artifact/io.github.zxing-cpp/android Thanks for all your input and help! Lets hope this will attract new users and make all the effort worth it. ;).

**Reply by axxel** [reply]
Question to all the Android folks (@markusfisch @ccrowell-kr @ISNing): Any idea why the most recent 2.3.0 [publishing](https://github.com/zxing-cpp/zxing-cpp/actions/runs/12579806400/job/35060612785) seemed to have [failed](https://central.sonatype.com/artifact/io.github.zxing-cpp/android)? Is this gradle warning the problem?

**Reply by ccrowell-kr** [reply]
It looks like the latest one succeeded, is it possible the artifact was [deployed but needs to be released to maven central](https://central.sonatype.org/publish/release/#releasing-deployment-from-ossrh-to-the-central-repository-introduction)?

**Reply by axxel** [reply]
What do you mean with "succeeded"? I can't find the 2.3.0 release anywhere. When I log into the Nexus managing console (https://s01.oss.sonatype.org) - which is highly unintuitive and slow by the way - I can't find a 2.3.0 artifact anywhere.

**Reply by ccrowell-kr** [reply]
Sorry for being ambiguous, I was saying the [pipeline job you linked](https://github.com/zxing-cpp/zxing-cpp/actions/runs/12579806400/job/35060612785) indicates success, so I would expect it to be available in the nexus console.

When 2.2.0 was published was it automatically 'released'?

**Reply by axxel** [reply]
> Sorry for being ambiguous, I was saying the [pipeline job you linked](https://github.com/zxing-cpp/zxing-cpp/actions/runs/12579806400/job/35060612785) indicates success, so I would expect it to be available in the nexus console.

So did I. ;)

> When 2.2.0 was published was it automatically 'released'?

Frankly, I don't remember. But it anyway, it certainly showed up "somewhere", while I can't find any hint at the 2.3.0 release now.
